ibis/backends/base/sql/alchemy/query_builder.py

```diff
@@ -344,39 +344,47 @@ def _convert_group_by(self, exprs):
 
 
 class AlchemySetOp(SetOp):
-    @classmethod
-    def reduce(cls, left, right, distincts):
-        distinct = next(distincts)
-        sa_func = cls.distinct_func if distinct else cls.non_distinct_func
-        return sa_func(left, right)
-
     def compile(self):
         context = self.context
         selects = []
 
+        def call(distinct, *args):
+            return (
+                self.distinct_func(*args)
+                if distinct
+                else self.non_distinct_func(*args)
+            )
+
         for table in self.tables:
             table_set = context.get_compiled_expr(table)
             selects.append(table_set.cte().select())
 
-        return functools.reduce(
-            functools.partial(self.reduce, distincts=iter(self.distincts)),
-            selects,
-        )
+        if len(set(self.distincts)) == 1:
+            # Use a single call when possible
+            return call(self.distincts[0], *selects)
```

Comment:
Does the `else` branch generalize to handle this case? If so, I'm inclined to drop the special case of a single `self.distincts` element.

Reply:
The `else` branch alone would work, but it would generate unnecessary subqueries leading to less concise SQL. I don't feel like this branch is overly complicated - I've updated the comments here to better clarify what's going on.

Reply:
Yep, didn't realize it would complicate the SQL. No need to do that!